verification/verify.go

```diff
@@ -112,7 +121,8 @@ func (sv *SegmentVerifier) Verify(params *Params) (*Params, error) {
 		if len(res.Pixels) != len(params.Results.Segments) {
 			// TODO make allowances for the verification algos not doing
 			//      pixel counts themselves; adapt broadcast.go verifyPixels
-			err = ErrPixelsAbsent
+
+			err = verifyPixelParams(params)
```

Comment:
Looking at this now I'm realizing that there could be an opportunity to simplify things a bit since the for loop below already performs the comparison of computed pixels vs. reported pixels for each of the rendition segments. At the moment, we're repeating this comparison each time we run `verifyPixelParams()`. It'd be nice if we can just count pixels when we need to and then rely on the below for loop for the comparison of computed pixels vs. reported pixels.

The below for loop relies on the `Pixels` field of the `Results` object returned from `policy.Verifier.Verify()`. We could create a type that implements the `Verifier` interface and its `Verify()` method it would be similar to the contents of the current `verifyPixelsParams()` function except instead of doing the comparison against reported pixels it would just populate the `Pixels` field of the returned `Results` object. This would be the "default" implementation of the `Verifier` interface (related to #1328) that would be used unless an external verifier is specified (would still need the on-chain check that we currently have in `verifyPixelsParams()`). WDYT?

Reply:
I think it all sounds nice, but it is scope creep.  Please write this into a new ticket so that we can close this one and demonstrate progress.

Reply:
I see your point about my suggestion representing additional work that is a bit different from the original suggested implementation approach in #1326. That being said, it represents an alternative approach to accomplishing the same objective that could be cleaner so I think its worthwhile having the discussion about whether we prefer the alternative approach or the current approach.

I brought it up because I think it could be a cleaner approach code-wise that doesn't seem to require a significant amount of additional effort so I wanted to see what you think! If we do prefer the alternative approach then I think it makes sense to address it now rather than committing code that is only then to be immediately changed afterwords.

Reply:
I'm not even sure I understand what you're asking for.   For that reason, alone, this should be a separate ticket.  What you're asking for seems to be a new feature / optimization that depends on this PR to be merged beforehand.  What if this optimization causes some problems in the future?  That doesn't mean the refactoring needs to be reverted - it meant the optimization itself needs debugging.   I don't mind doing the work, but it *is* scope creep and I would prefer to see it on a new ticket.

Reply:
I'm not even sure I understand what you're asking for. For that reason, alone, this should be a separate ticket. What you're asking for seems to be a new feature / optimization that depends on this PR to be merged beforehand. What if this optimization causes some problems in the future? The entire refactoring should not be put into question - they are separate changes. I don't mind doing the work, but it is scope creep and I would prefer to see it on a new ticket.

Reply:
I think I see what you're talking about now, but it sounds like some heavy lifting.  I think this should be merged as-is first, if it is correct.  This way, we can introduce this first as a stable change before introducing new optimizations.

Reply:
Wouldn't we still need a default verifier in order to actually invoke this pixel checking? (Presumably the default verifier would be instantiated in onchain mode if the epic classifier wasn't).

But I'm indifferent as to whether the verification should remain here or moved within the default verifier. @yondon suggestion makes sense ; keeping it here would mitigate the requirement to re-implement pixel checking in future verifiers (although probably still desirable to do as a performance optimization).

Reply:
Here is a sketch of what I was thinking.

The default verifier would look something like this:

```
type DefaultVerifier struct {}

func (v *DefaultVerifier) Verify() (*Results, error) {
    return &Results{Score: 1.0}, nil
}
```
and `SegmentVerifier.Verify()` would do something like this:

```
res, err := sv.policy.Verifier.Verify(params)

if (...) {
    if len(res.Pixels) != len(params.Results.Segments) {
        if ( // on-chain check here ) {
            res.Pixels, err = countSegmentPixels(params.Results.Segments, params.Renditions)
        }
    }
    // Compare res.Pixels against reported pixels for each rendition here
}
```

The default verifier basically does nothing and the built-in accommodation for verifiers that don't implement pixel counting in `SegmentVerifier` would be executed when the node is in on-chain mode.

Reply:
@mk-livepeer Does the above comment clear things up a bit? If you still prefer to stick with the current approach and explore the approach described in the above comment separately then that is fine by me. Although keep in mind the approach we take here will affect how we handle #1328

Reply:
Thanks for the explanation @yondonfu - that helped me understand it better. So I guess we'd be talking about incorporating two distinct code paths here:

* The `verifyPixelParams` as implemented here which also does the pixel counting / checking . This would be run if the policy does *not* specify a verifier
* The proposed `countSegmentPixels` which only does the pixel counting. The default policy *could* define a stub verifier to leverage this.

Perhaps one way to unify the two would be to do the following:

```golang
if sv.policy.Verifier == nil {
  res = &Results{ Score: 1.0 }
} else {
  res, err = sv.policy.Verifier.Verify(params)
}

if (...) {
    if len(res.Pixels) != len(params.Results.Segments) {
        res.Pixels, err = countSegmentPixels(params.Results.Segments, params.Renditions)
    }
    // Compare res.Pixels against reported pixels for each rendition here
}
```

This is nice because it avoids having to define a struct for a stub verifier, and keeps the overall flow consistent with the only difference being the pixels are explicitly counted if the verifier doesn't do it.

Reply:
> If you prefer to stick with the current approach and explore the approach described in the above comment separately as an optimization then that is fine by me and I'll leave it up to you.

Yes, @yondonfu , I would greatly prefer to work on this optimization in a separate ticket.  If all else looks good in this PR, please approve.

Reply:
My feeling is this is still in scope - the `verifyPixelParams` is a new function on top of the refactor. Yondon's [suggestion](https://github.com/livepeer/go-livepeer/pull/1360#discussion_r376145833) combined with the [refinement](https://github.com/livepeer/go-livepeer/pull/1360#discussion_r376238249) means we end up with better code that's both simpler and more straightforward: two invocations of `verifyPixelParams` (count + check) can be replaced with a single invocation of `countPixels` (count-only) , and we reuse the existing pixel-checking code. Would rather do this now while we're working in this area; deferring it just increases friction and code churn.

Reply:
It feels like too many moving parts.  I'm attempting to address today's comments, but each small nit results in breaking half the tests.  How about a compromise?  Let's make a `staging` branch, as I've proposed in the past.  We can merge this to staging once all points are addressed besides this one, and then it can unblock my work on #1325 which depends on the merge of this PR.  Not to mention the additional work in #1328   ...   So I'm proposing that we fork master to a staging (or dev) branch and I rebase my branches-in-progress off that.  After the merge, the delta will be greatly reduced, there will be less moving parts, and it will be a lot easier to see what Yondon is proposing here.

Reply:
Thought about this some more and I do think that it would be good idea to update the current implementation given that Josh's [refined approach](https://github.com/livepeer/go-livepeer/pull/1360#discussion_r376238249) should result in less code and less complexity which should make any additional work we have do in this area of the codebase easier.

> It feels like too many moving parts.

I think the update should result in less moving parts in this PR since there will be less code overall and fewer code paths that a reader needs to follow.

> and then it can unblock my work on #1325

If code in this PR is a dependency for #1325 we can always branch off this PR. I see your point about merging PRs to a `staging` branch that contain code changes in an intermediate state (i.e. there will be a follow up PR to bring the code changes to a finished state), but I think that can be a discussion separate from the issue at hand which is the desire to update the implementation in this PR given that a simpler alternative is available.

> Not to mention the additional work in #1328

Josh makes a good point in [this comment](https://github.com/livepeer/go-livepeer/pull/1360#commitcomment-37177513) that it would make sense to address #1328 in this PR given that how we structure this PR will impact how we address that issue.

Reply:
how's this?  refactored in 6aa915f5d7256007052c99752b38355cb0f8bab7

Reply:
now integrated the solution for #1328